wdl-cli/src/analysis.rs

```diff
@@ -0,0 +1,202 @@
+        self.exceptions.extend(rules);
+        self
+    }
+
+    /// Sets whether linting is enabled.
+    pub fn lint(mut self, value: bool) -> Self {
+        self.lint = value;
+        self
+    }
+
+    /// Sets whether `shellcheck` is enabled.
+    pub fn shellcheck(mut self, value: bool) -> Self {
+        self.shellcheck = value;
+        self
+    }
+
+    /// Sets the initialization callback.
+    pub fn init<F>(mut self, init: F) -> Self
+    where
+        F: Fn() + 'static,
+    {
+        self.init = Box::new(init);
+        self
+    }
+
+    /// Sets the progress callback.
+    pub fn progress<F>(mut self, progress: F) -> Self
+    where
+        F: Fn(ProgressKind, usize, usize) -> BoxFuture<'static, ()> + Send + Sync + 'static,
+    {
+        self.progress = Box::new(progress);
+        self
+    }
+
+    /// Runs the analysis and returns all results (if any exist).
+    pub async fn run(self) -> std::result::Result<AnalysisResults, NonEmpty<Arc<Error>>> {
+        warn_unknown_rules(&self.exceptions);
+        let config = get_diagnostics_config(&self.exceptions);
+
+        (self.init)();
+
+        let validator = Box::new(move || {
+            let mut validator = Validator::default();
+
+            if self.lint {
+                let visitor = get_lint_visitor(&self.exceptions);
+                validator.add_visitor(visitor);
+            }
+
+            validator
+        });
+
+        let mut analyzer = Analyzer::new_with_validator(
+            config,
+            move |_, kind, count, total| (self.progress)(kind, count, total),
+            validator,
+        );
+
+        for source in self.sources {
+            if let Err(error) = source.register(&mut analyzer).await {
+                return Err(NonEmpty::new(Arc::new(error)));
+            }
+        }
+
+        let results = analyzer
+            .analyze(())
+            .await
+            .map_err(|error| NonEmpty::new(Arc::new(error)))?;
+
+        AnalysisResults::try_new(results)
+    }
+}
+
+impl Default for Analysis {
+    fn default() -> Self {
+        Self {
+            sources: Default::default(),
+            exceptions: Default::default(),
+            lint: Default::default(),
+            shellcheck: Default::default(),
+            init: Box::new(|| {}),
+            progress: Box::new(|_, _, _| Box::pin(async {})),
+        }
+    }
+}
+
+/// Warns about any unknown rules.
+fn warn_unknown_rules(exceptions: &HashSet<String>) {
+    let mut names = wdl_analysis::rules()
+        .iter()
+        .map(|rule| rule.id().to_owned())
+        .collect::<Vec<_>>();
+
+    names.extend(wdl_lint::rules().iter().map(|rule| rule.id().to_owned()));
+
+    let mut unknown = exceptions
+        .iter()
+        .filter(|rule| !names.iter().any(|name| name.eq_ignore_ascii_case(rule)))
+        .map(|rule| format!("`{rule}`"))
+        .collect::<Vec<_>>();
+
+    if !unknown.is_empty() {
+        unknown.sort();
+
+        warn!(
```

Comment:
This seems fine for `wdl-cli`, but in the sprocket bin we probably want to use `find_nearest_rule()` ([here](https://github.com/stjude-rust-labs/wdl/blob/main/wdl-lint/src/util.rs#L108)) to offer a suggestion of what they might have meant to type.
Just something to keep in mind while designing this API

Reply:
Okay, I will incorporate that into my Sprocket PR when I get around to working on that one again.